lisa/operating_system.py

```diff
@@ -179,7 +245,50 @@ def install_packages(
         if self._first_time_installation:
             self._first_time_installation = False
             self._initialize_package_installation()
-        self._install_packages(package_names)
+
+        self._install_packages(package_names, signed)
+
+    # query_package
+    # Query if a package/tool is installed on the node.
+    # Return Value - bool
+    def query_package(
+        self, package: Union[str, Tool, Type[Tool]], signed: bool = True
+    ) -> bool:
+        if isinstance(package, str):
+            package_name = package
+        elif isinstance(package, Tool):
+            package_name = package.package_name
+        else:
+            assert isinstance(package, type), f"actual:{type(package)}"
+            # Create a temp object, it doesn't query.
+            # So they can be queried together.
+            tool = package.create(self._node)
+            package_name = tool.package_name
+
+        return self._query_package(package_name)
+
+    # TODO: Implement update_packages
+    def update_packages(
+        self, packages: Union[str, Tool, Type[Tool], List[Union[str, Tool, Type[Tool]]]]
+    ) -> None:
+        pass
+
+    def _get_os_version(self) -> OsVersion:
+        cmd_result = self._node.execute(cmd="cat /etc/os-release", no_error_log=True)
```

Comment:
This file may not exist, the implementation should be in it's derived distros, not in Posix class.

Reply:
yup, forgot to add the else part. It should call the underlying distro's method in else part

Reply:
I believe many newer distros have this now, so its no harm to try and get the info from this file, if it doesnt exist then it can go and check in the specific file of distro

Reply:
1. The `Posix` class is for both Linux and BSD, I'm not sure how it's supported in BSD. At least, it should be moved to Linux class.
2. See the `_get_detect_string` method. It's tested on 12k images in Azure. After `cat /etc/os-release`, they are extra commands we need to run to support other distros. If it's the base implementation, it should throw exception, if this command doesn't success. So we know which one is not supported, and implement in subclasses.

Reply:
And if there is some old distro we need to continue to support, we need to handle them.

Reply:
> And if there is some old distro we need to continue to support, we need to handle them.

Old distros are supported, the else part will take care of it which will take it to specific distro's get_os_version implementation

Reply:
> 1. The `Posix` class is for both Linux and BSD, I'm not sure how it's supported in BSD. At least, it should be moved to Linux class.
> 2. See the `_get_detect_string` method. It's tested on 12k images in Azure. After `cat /etc/os-release`, they are extra commands we need to run to support other distros. If it's the base implementation, it should throw exception, if this command doesn't success. So we know which one is not supported, and implement in subclasses.

I believe it is in FreeBSD now
https://reviews.freebsd.org/D22271

Reply:
There should be an `else` block for `if cmd_result.exit_code == 0 and cmd_result.stdout != "":`. If the condition doesn't meet, raise an exception. So we can find which one is not supported.

Reply:
there is an else block now